### PR TITLE
Add tree component

### LINKED ISSUE #590: Tree component
This is an API proposal for the tree component per the following sketch:

image

### Design goals

* component renders elements in a folder tree-like structure, rendering a label and an optional `NameTag` at each node. The component should be agnostic about the level of nesting to avoid verbose type definitions.
* supports any HTML attributes and event listeners at each level - both click events as well as drag-and-drop handlers are important for individual nodes.
* the outside world need not know about open-closed state - this should be handled in internal state.
* opinionated on node styling (font-weight, color etc.) based on the nesting level / whether there is a tag or not.

### API

The basis for the props of this component is a recursive data structure that models an arbitrarily nested tree:

```
export interface Tree {
  label: string
  tag?: string
  childNodes: Tree[]
}
```

Create a component like so:

```
(
  <Tree
    tree={{
      label: "Dimensions",
      childNodes: [
        {
          label: "Dim 1",
          tag: "DM",
          childNodes: [],
        },
        {
          label: "Dim 2",
          tag: "DM",
          onClick: () => {
            console.log("clicked")
          },
          childNodes: [
            {
              label: "Subdim 1",
              tag: "SM",
              childNodes: [],
            },
          ],
        },
      ],
    }}
  />
)
```

The component's implementation handles active state by keeping track of active node paths in an array, sorting out conflicts when new props arrive (although this should be infrequent).

### PR DESCRIPTION
**Pull request by peterszerzo**
### Summary

Collapsible tree component per following gif:

tree3

### Progress

* [x] basic state management, public API.
* [x] handle initially open state.
* [x] handle custom tag colors.
* [x] handle disabled state.
* [x] visual fine-tuning.

### Related issue

Fixes #590.

https://trello.com/c/hAECaWyR/178-simple-expandable-tree-with-3-levels

### To be tested

Tester 1 - Tekas

- [x] No error/warning in the console
- [x] Looks and works as expected with tagged nodes at different levels.

Tester 2

- [ ] No error/warning in the console
- [ ] Looks and works as expected with tagged nodes at different levels.

### CODE REVIEW

**Comment by TejasQ** [comment]
The code looks incredible so far!

**Comment by TejasQ** [comment]
[Simple expandable tree with 3 levels](https://trello.com/c/hAECaWyR/178-simple-expandable-tree-with-3-levels)

**Comment by TejasQ** [comment]
This fixes #590

**Comment by micha-f** [comment]
@peterszerzo Looks great! Would you be ok with making the colored tags just a little bit smaller?

**Comment by peterszerzo** [comment]
@micha-f I agree - a bit of a shrink on top and bottom padding would do it a lot of good visually. Since we are using the `NameTag` component internally, can we make that change across all UI's?

**Comment by micha-f** [comment]
I am not sure whether this needs to use the name tag. Originally it was supposed to look quite different. More like a square with rounded corners - and only a single letter inside.

image

File: packages/components/src/Tree/Tree.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+              />
+            ))}
+          </TreeChildren>
+        )}
+    </TreeContainer>
+  )
+}
+
+class Tree extends React.Component<Props, State> {
+  state: State = {
```
Comment (TejasQ) [inline packages/components/src/Tree/Tree.tsx]:
```diff
- state: State = {
+ readonly state: State = {
```

since state is readonly and it's more explicit.

File: packages/components/src/Tree/Tree.tsx [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                openPaths={openPaths}
+              />
+            ))}
+          </TreeChildren>
+        )}
+    </TreeContainer>
+  )
+}
+
+class Tree extends React.Component<Props, State> {
```
Comment (TejasQ) [inline packages/components/src/Tree/Tree.tsx]: You could also

```diff
- class Tree extends React.Component<Props, State> {
+ class Tree extends React.Component<Props, Readonly<State>> {
```

**Comment by TejasQ** [comment]
It may be helpful to add `user-select: none` because I accidentally highlight stuff while clicking on nodes.

**Comment by peterszerzo** [comment]
@micha-f makes sense - I'll take care of adding the square!

**Comment by TejasQ** [comment]
@peterszerzo does this support drag and drop?

**Comment by peterszerzo** [comment]
@TejasQ - yepp, have a look at the code snippet in netlify - one of the nodes has the appropriate props so that it's draggable.

**Comment by peterszerzo** [comment]
@micha-f I updated the design - see updated gif above and the netlify link.

**Comment by micha-f** [comment]
Looks great! The single letter ("D") feels just a little bit too big for me. Maybe we can increase the font-size by one?

File: packages/components/src/Tree/Tree.utils.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * Adds or removes a path from a list of paths.
+ * togglePath([1, 2])([[1, 2, 3], [0]]) -> [[1, 2, 3], [0], [1, 2]]
+ * togglePath([1, 2])([[1, 2], [0]]) -> [[0]]
+ */
+export const togglePath = (path: number[]) => (paths: number[][]): number[][] => {
+  if (paths.length === 0) {
+    return [path]
+  }
+  const [head, ...tail] = paths
+  if (head.join("") === path.join("")) {
```
Comment (TejasQ) [inline packages/components/src/Tree/Tree.utils.ts]: What is this supposed to do?

File: packages/components/src/Tree/Tree.utils.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  }
+  return [head, ...togglePath(path)(tail)]
+}
+
+export const containsPath = (path: number[]) => (paths: number[][]): boolean => {
+  if (paths.length === 0) {
+    return false
+  }
+  const [head, ...tail] = paths
+  if (head.join("") === path.join("")) {
```
Comment (TejasQ) [inline packages/components/src/Tree/Tree.utils.ts]: You seem to have a bit of duplication here (see line 22).

File: packages/components/src/Tree/Tree.utils.ts (hunk 3 above)
Comment (fabien0102) [reply packages/components/src/Tree/Tree.utils.ts]: I bet on an Array compare :-p

Comment (peterszerzo) [reply packages/components/src/Tree/Tree.utils.ts]: It checks whether a path is the same as another path like [ 0, 1 ] with another [ 0, 1 ]

File: packages/components/src/Tree/Tree.types.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+/**
+ * ITree naming convention is used because it would otherwise clash with the name of the component.
```
Comment (TejasQ) [inline packages/components/src/Tree/Tree.types.ts]: This isn't really needed is it?

**Pull request merged**